src/net/server/connection.rs

```diff
@@ -0,0 +1,633 @@
+
+        self.transceive_until(
+            command_rx,
+            state,
+            stream_rx,
+            result_q_rx,
+            &mut msg_buf,
+        )
+        .await?;
+
+        state.full_msg_received();
+
+        self.process_request(
+            msg_buf,
+            self.addr,
+            state.result_q_tx.clone(),
+            self.middleware_chain.clone(),
+            &self.service,
+            self.metrics.clone(),
+        )
+        .map_err(ConnectionEvent::ServiceError)?;
+
+        Ok(())
+    }
+
+    async fn transceive_until(
+        &self,
+        command_rx: &mut watch::Receiver<ServiceCommand>,
+        state: &mut StreamState<Stream, Buf, Svc>,
+        stream_rx: &mut ReadHalf<Stream>,
+        result_q_rx: &mut mpsc::Receiver<CallResult<Svc::Target>>,
+        buf: &mut <Buf as BufSource>::Output,
+    ) -> Result<ConnectionEvent<Svc::Error>, ConnectionEvent<Svc::Error>>
+    {
+        // Note: The MPSC receiver used to receive finished service call
+        // results can be read from safely even if the future is cancelled.
+        // Thus we don't need to keep the future, we can just call recv()
+        // again when we need to.
+        //
+        // The same is not true of reading an exact number of bytes from the
+        // incoming data stream, the future cannot be cancelled safely as any
+        // bytes already read will be written to the buffer but we will lose
+        // the knowledge of how many bytes have been written to the buffer. So
+        // we must keep using the same future until it finally resolves when
+        // the read is complete or results in an error.
+        'read: loop {
+            let stream_read_fut = stream_rx.read_exact(buf.as_mut());
+            let timeout_fut = tokio::time::sleep(state.timeout_as_std());
+
+            tokio::pin!(stream_read_fut);
+            tokio::pin!(timeout_fut);
+
+            loop {
+                tokio::select! {
+                    biased;
+
+                    res = command_rx.changed() => {
+                        self.process_service_command(res, state, command_rx)?;
+                    }
+
+                    result_q_res = result_q_rx.recv() => {
+                        // If we failed to read the results of requests
+                        // processed by the service because the queue holding
+                        // those results is empty and can no longer be read
+                        // from, then there is no point continuing to read
+                        // from the input stream because we will not be able
+                        // to access the result of processing the request.
+                        // TODO: Describe when this can occur.
+                        let call_result = result_q_res
+                            .ok_or(ConnectionEvent::DisconnectWithFlush)?;
+
+                        self.process_queued_result(state, call_result).await;
```

Comment:
I might be misunderstanding how this works but looking at the `tokio::select!` docs it says:

Given:
```
<pattern> = <async expression> (, if <precondition>)? => <handler>,
```

> The macro aggregates all <async expression> expressions and runs them concurrently on the *current* task. Once the *first* expression completes with a value that matches its <pattern>, the select! macro returns the result of evaluating the completed branch’s <handler> expression.

And:

> This behavior can be overridden by adding biased; to the beginning of the macro usage. See the examples for details. This will cause select to poll the futures in the order they appear from top to bottom. 

Now lets walk through the lifecycle:

1. _Evaluate all `<precondition>`s first_. We don't have any of those, so let's move on.
2. _Aggregate the `<async expression>`s from each branch_. Ok.
3. _Concurrently await on the results for all remaining `<async expressoin>`s_. Ok.
4. _Once an `<async expression>` returns a value, attempt to apply the value to the provided `<pattern>`, if the pattern matches, evaluate `<handler>` and return_. So if the `recv()` expresson returns a value, the `<pattern>` will match, and the `<handler>` will be evaluated. Based on the other doc comment above, as this is the _*first* expression to complete with a value_ I understand that the future for the `changed()` branch will already have been dropped because `biased` causes polling in sequence, but even if had it hadn't been dropped already, recv() is cancel safe. We will then invoke `process_queued_result` and `await` the result, and until that completes the select block won't do anything else and won't run again until the loop goes round again. If the `&mut stream_read_fut` future had been polled that is NOT cancel safe, but it won't be polled until after polling `recv()` completes, so we don't get that far. Likewise `timeout_fut` isn't polled but even if it was that is cancel safe anyway.

Am I missing something?

Reply:
I think the `stream_read_fut` part isn't cancel safe because it is a call to `read_exact()` which is NOT cancel safe, and if the `recv()` future completes while the `read_exact()` future is running any data read by `read_exact()` will be lost.

This wasn't an issue with the code I had when I first wrote this PR as that used `futures::future::select()` which returns the incomplete future rather than dropping it.

However it may still be possible with `tokio::select!()` to make that safe if a `&mut` future can be used, per: https://users.rust-lang.org/t/tokio-select-without-cancellation/71300/6 (which I can't see in the [documentation](https://docs.rs/tokio/latest/tokio/macro.select.html).